### DISCUSSION TITLE
How to store provider_refresh_token during OAuth flow

### DISCUSSION THREAD

**Discussion by sp88011** [supabase/supabase#22653] [Questions] [answered]
Hi! How would I go about storing the `provider_refresh_token` in a table (separate from `auth.users`) when using Google OAuth?

I'd love to be able to store it using a database trigger just [like how the official docs explain storing user data upon sign-up](https://supabase.com/docs/guides/auth/managing-user-data#advanced-techniques) but it seems like the `provider_refresh_token` is not accessible at that point?

Right now, I'm doing a `getSession()` call inside the `/callback` route...while this works, it feels wrong.

What do you guys do or recommend... Keep as is? Do something else?

Current implementation:
`/api/callback`
```
import { db } from "@/drizzle/db";
import { userMeta } from "@/drizzle/schema";
import { createSupabaseServerClient } from "@/supabase/server";
import { getURL } from "@/utils/utils";
import { eq } from "drizzle-orm";
import { revalidatePath } from "next/cache";
import { NextResponse } from "next/server";

export async function GET(request: Request) {
  try {

    const requestUrl = new URL(request.url);
    const code = requestUrl.searchParams.get("code");

    if (code) {
      const supabase = await createSupabaseServerClient();

      await supabase.auth.exchangeCodeForSession(code);

      //persist provider_refresh_token
      const session = await supabase.auth.getSession();

      if (session?.data) {
        const userId = session.data.session?.user.id;
        const providerRefreshToken = session?.data.session
          ?.provider_refresh_token;
        const providerToken = session?.data.session?.provider_token;

        if (providerRefreshToken && userId) {
          await db.update(userMeta).set({ //store token in database
            providerRefreshToken,
            providerToken,
          }).where(eq(userMeta.id, userId));
        }
      }
    }

    // URL to redirect to after sign up process completes
    return NextResponse.redirect(`${getURL()}/home/`);
  } catch (e) {
    console.log(e);
    return NextResponse.redirect(`${getURL()}/login`);
  }
}

```

Here's what the [docs](https://supabase.com/docs/guides/auth/social-login/auth-google#using-the-oauth-flow-for-web) say about the `provider_refresh_token` (no word about how to store it):

> 
> You can additionally extract the
> provider_token
> from the session (on initial login only) which is the OAuth 2.0 access token issued by Google that grants your application access to the Google services for the authenticated users. Please store this token in local storage, cookies or in your database or server.
> 
> Google does not send out a refresh token by default, so you will need to pass parameters like these to
> signInWithOAuth()
> in order to extract the
> provider_refresh_token
> :
> 
> const { data, error } = await supabase.auth.signInWithOAuth({
>   provider: 'google',
>   options: {
>     queryParams: {
>       access_type: 'offline',
>       prompt: 'consent',
>     },
>   },
> })
> 
>

**Comment by GaryAustin1** [comment, marked as answer]
A bit more in the doc for the call:
https://supabase.com/docs/reference/javascript/auth-signinwithoauth

**Reply by sp88011** [reply]
thanks, great detective work!
